File: bench/render/raylib.ts
import {DrawScene, camera} from './geometry.js';
import {Overlay,palette,markerTriangles} from './overlay.js';
import type {SnapshotCopy} from '../../wasm/index.js';
interface RaylibModule {
  HEAPF32: Float32Array; HEAPU32: Uint32Array;
  _sl_render_init(instances:number,vertices:number,batches:number,width:number,height:number):number;
  _sl_render_vertices():number; _sl_render_poses():number; _sl_render_batches():number;
  _sl_render_overlay_init(lines:number,markers:number):number;
  _sl_render_overlay_lines():number;_sl_render_overlay_markers():number;_sl_render_overlay_colors():number;
  _sl_render_overlay_palette():number;_sl_render_overlay_mesh():number;
  _sl_render_overlay_counts(lines:number,markers:number):number;
  _sl_render_study_poses(study:number,poses:number,count:number):number;
  _sl_render_overlay_refresh(study:number,scale:number,x:number,y:number):number;
  _sl_render_overlay_line_count():number;_sl_render_overlay_marker_count():number;
  _sl_render_draw(scale:number,x:number,y:number):number; _sl_render_dispose():void;
}
export async function createRaylibCandidate(canvas:HTMLCanvasElement, scene:DrawScene,overlay?:Overlay) {
  if (!canvas.id || document.getElementById(canvas.id)!==canvas) throw new Error('Raylib requires an attached canvas with a unique id');
  const url=new URL('./raylib.mjs',import.meta.url).href;
  const {default:factory}=await import(url);
  const m:RaylibModule=await factory({canvas,wasmMemory:new WebAssembly.Memory({initial:1024,maximum:1024})});
  return initializeCandidate(m,canvas,scene,overlay);
}
interface DirectSource {study:number;steps():number;valid():void;diagnostic:boolean;
  bodyCapacity:number;contactCapacity:number;jointCapacity:number}
function initializeCandidate(m:RaylibModule,canvas:HTMLCanvasElement,scene:DrawScene,overlay?:Overlay,source?:DirectSource,
  registerRebind?:(rebind:(next:DirectSource)=>void)=>void) {
  const vertexCount=scene.meshes.reduce((n,mesh)=>n+mesh.triangles.length/2,0);
  if (!m._sl_render_init(scene.count,vertexCount,scene.batches.length,canvas.width,canvas.height))
    throw new Error('Raylib renderer initialization failed');
  try {
    const vertices=new Float32Array(m.HEAPF32.buffer,m._sl_render_vertices(),2*vertexCount);
    const poses=new Float32Array(m.HEAPF32.buffer,m._sl_render_poses(),4*scene.count);
    const batches=new Uint32Array(m.HEAPU32.buffer,m._sl_render_batches(),5*scene.batches.length);
    const offsets:number[]=[]; let offset=0;
    for (const mesh of scene.meshes) { offsets.push(offset/2); vertices.set(mesh.triangles,offset); offset+=mesh.triangles.length; }
    scene.batches.forEach((b,i)=>batches.set([b.first,b.count,offsets[b.mesh],scene.meshes[b.mesh].triangles.length/2,b.color],i*5));
    const view=camera(scene.rect,canvas.width,canvas.height);
    let lines:Float32Array|undefined,markers:Float32Array|undefined,colors:Float32Array|undefined;
    if(overlay||source?.diagnostic){
      const lineCapacity=overlay?overlay.lines.length/5:4*source!.bodyCapacity+2*source!.contactCapacity+source!.jointCapacity+6;
      const markerCapacity=overlay?overlay.markers.length/3:2*source!.contactCapacity+2*source!.jointCapacity+2;
      if(!m._sl_render_overlay_init(lineCapacity,markerCapacity))throw new Error('Raylib diagnostic allocation failed');
      if(overlay){
        lines=new Float32Array(m.HEAPF32.buffer,m._sl_render_overlay_lines(),overlay.lines.length);
        markers=new Float32Array(m.HEAPF32.buffer,m._sl_render_overlay_markers(),overlay.markers.length);
        colors=new Float32Array(m.HEAPF32.buffer,m._sl_render_overlay_colors(),overlay.bodyColors.length);
      }
      new Float32Array(m.HEAPF32.buffer,m._sl_render_overlay_mesh(),markerTriangles.length).set(markerTriangles);
      new Uint32Array(m.HEAPU32.buffer,m._sl_render_overlay_palette(),48).set(palette.flatMap(hex=>[1,3,5].map(at=>parseInt(hex.slice(at,at+2),16))));
    }
    let revision=0,overlayRevision=0,disposed=false,lastStep=-1;
    registerRebind?.(next=>{
      if(disposed||!source||next.diagnostic!==source.diagnostic||next.bodyCapacity!==source.bodyCapacity||
        next.contactCapacity!==source.contactCapacity||next.jointCapacity!==source.jointCapacity)throw new Error('Incompatible renderer transfer');
      next.valid();source=next;lastStep=-1;
    });
    return {
      // CPU-to-WASM pose copies are separate from raylib's internal GPU uploads.
      poseCopyBytes:0,posePrepareBytes:0,diagnosticCopyBytes:0,diagnosticPrepareBytes:0,lineCount:0,markerCount:0,
      uploadBytes:null, gpuBytes:null, drawCalls:null,
      linearMemoryBytes:m.HEAPF32.buffer.byteLength,
      draw() {
        if (disposed) throw new Error('Disposed raylib candidate');
        this.poseCopyBytes=0;this.posePrepareBytes=0;this.diagnosticCopyBytes=0;this.diagnosticPrepareBytes=0;
        if(source){
          source.valid();const step=source.steps();
          if(lastStep!==step){
            if(!m._sl_render_study_poses(source.study,poses.byteOffset,scene.count))throw new Error('Direct C pose preparation failed');
            if(source.diagnostic){
              if(!m._sl_render_overlay_refresh(source.study,view.scale,view.x,view.y))throw new Error('Direct C diagnostic preparation failed');
              this.lineCount=m._sl_render_overlay_line_count();this.markerCount=m._sl_render_overlay_marker_count();
              this.diagnosticPrepareBytes=20*this.lineCount+12*this.markerCount+4*scene.count;
            }
            lastStep=step;this.posePrepareBytes=poses.byteLength;
          }
        }else if (revision!==scene.revision) { poses.set(scene.poses); revision=scene.revision; this.poseCopyBytes=poses.byteLength; }
        if(overlay&&lines&&markers&&colors&&overlayRevision!==overlay.revision){
          // Active prefixes only, without temporary typed-array views. The
          // co-located C study will remove this prototype boundary copy.
          for(let i=0;i<overlay.lineCount*5;++i)lines[i]=overlay.lines[i];
          for(let i=0;i<overlay.markerCount*3;++i)markers[i]=overlay.markers[i];
          colors.set(overlay.bodyColors);
          if(!m._sl_render_overlay_counts(overlay.lineCount,overlay.markerCount))throw new Error('Invalid raylib diagnostic commands');
          this.diagnosticCopyBytes=overlay.lineCount*20+overlay.markerCount*12+overlay.bodyColors.byteLength;
          this.lineCount=overlay.lineCount;this.markerCount=overlay.markerCount;
          overlayRevision=overlay.revision;
        }
        if (!m._sl_render_draw(view.scale,view.x,view.y)) throw new Error('Raylib draw failed');
      },
      dispose() { if (!disposed) { m._sl_render_dispose(); disposed=true; } }
    };
  } catch (error) { m._sl_render_dispose(); throw error; }
}

interface StudyWorld {
  configuration:{scene:'pyramid'|'rain'|'chains';copies:number;bodyCapacity:number;contactCapacity:number;jointCapacity:number};
  snapshot:SnapshotCopy;steps:number;refreshSnapshot(diagnostic?:boolean):boolean;
}
/** One fixed memory owns physics and raylib. Geometry/palette cross JS during
 * setup; live poses and diagnostic commands are prepared directly in C. */
export async function createRaylibStudyModule(canvas:HTMLCanvasElement,{memoryBytes=64*1024*1024}={}) {
  if(!canvas.id||document.getElementById(canvas.id)!==canvas)throw new Error('Raylib requires an attached canvas with a unique id');
  if(!Number.isInteger(memoryBytes)||memoryBytes<64*1024*1024||memoryBytes>512*1024*1024||memoryBytes%65536)
    throw new RangeError('Invalid fixed raylib memory budget');
  const {default:factory}=await import(new URL('./raylib.mjs',import.meta.url).href);
  const {ownStudyModule}=await import(new URL('./physics/owner.mjs',import.meta.url).href);
  const m:RaylibModule=await factory({canvas,wasmMemory:new WebAssembly.Memory({initial:memoryBytes/65536,maximum:memoryBytes/65536})});
  const rebinders=new WeakMap<object,{diagnostic:boolean;rebind:(next:DirectSource)=>void}>();
  return ownStudyModule(m,memoryBytes,(study:number,world:StudyWorld,valid:()=>void,options:{diagnostic?:boolean},previous?:object)=>{
    if(!options||typeof options!=='object'||Array.isArray(options)||
        (options.diagnostic!==undefined&&typeof options.diagnostic!=='boolean'))throw new TypeError('Invalid renderer options');
    const existing=previous?rebinders.get(previous):undefined;
    if(previous&&!existing)throw new Error('Renderer belongs to another module');
    const diagnostic=existing?.diagnostic??options.diagnostic??false;
    const source={study,valid,steps:()=>world.steps,diagnostic,bodyCapacity:world.configuration.bodyCapacity,
      contactCapacity:world.configuration.contactCapacity,jointCapacity:world.configuration.jointCapacity};
    if(existing){existing.rebind(source);return previous;}
    if(!world.refreshSnapshot())throw new Error('Raylib geometry snapshot failed');
    const scene=new DrawScene(world.snapshot,world.configuration.scene,undefined,world.configuration.copies);
    let rebind!:(next:DirectSource)=>void;
    const renderer=initializeCandidate(m,canvas,scene,undefined,source,value=>{rebind=value;});
    rebinders.set(renderer,{diagnostic,rebind});return renderer;
  });
}

File: bench/render/raylib.c
#include "raylib.h"
#include <math.h>
#include <raylib.h>
#include <rlgl.h>
#include <stdlib.h>

static float *vertices;
static float *poses;
static uint32_t *batches;
static uint32_t instance_count;
static uint32_t vertex_count;
static uint32_t batch_count;
static bool ready;
static float *overlay_lines, *overlay_markers, *overlay_colors;
static float overlay_mesh[180];
static uint32_t overlay_palette[48];
static uint32_t line_capacity, marker_capacity, line_count, marker_count;

static void overlay_color(float code)
{
    const uint32_t at = (uint32_t)code * 3u;
    rlColor4ub((unsigned char)overlay_palette[at],
               (unsigned char)overlay_palette[at + 1u],
               (unsigned char)overlay_palette[at + 2u], 255u);
}
static void overlay_draw(void)
{
    rlBegin(RL_TRIANGLES);
    for (uint32_t i = 0u; i < line_count; ++i) {
        const float *p = overlay_lines + 5u * i;
        const float dx = p[2] - p[0], dy = p[3] - p[1];
        const float length_sq = dx * dx + dy * dy;
        const float inverse = length_sq > 0.0f ? 0.5f / sqrtf(length_sq) : 0.0f;
        const float nx = -dy * inverse, ny = dx * inverse;
        overlay_color(p[4]);
        rlVertex2f(p[0] - nx, p[1] - ny);
        rlVertex2f(p[2] - nx, p[3] - ny);
        rlVertex2f(p[2] + nx, p[3] + ny);
        rlVertex2f(p[0] - nx, p[1] - ny);
        rlVertex2f(p[2] + nx, p[3] + ny);
        rlVertex2f(p[0] + nx, p[1] + ny);
    }
    for (uint32_t i = 0u; i < marker_count; ++i) {
        const float *p = overlay_markers + 3u * i;
        overlay_color(p[2]);
        for (uint32_t v = 0u; v < 90u; ++v) {
            rlVertex2f(p[0] + 3.0f * overlay_mesh[2u * v],
                       p[1] + 3.0f * overlay_mesh[2u * v + 1u]);
        }
    }
    rlEnd();
}

bool sl_render_init(uint32_t instances, uint32_t vertices_count,
                    uint32_t batches_count, uint32_t width, uint32_t height)
{
    if (ready || instances == 0u || instances > 65536u ||
        vertices_count == 0u || vertices_count > 65536u * 90u ||
        batches_count == 0u || batches_count > instances || width == 0u ||
        height == 0u || width > 8192u || height > 8192u) {
        return false;
    }
    vertices = calloc((size_t)vertices_count * 2u, sizeof(float));
    poses = calloc((size_t)instances * 4u, sizeof(float));
    batches = calloc((size_t)batches_count * 5u, sizeof(uint32_t));
    if (vertices == NULL || poses == NULL || batches == NULL) {
        sl_render_dispose();
        return false;
    }
    instance_count = instances;
    vertex_count = vertices_count;
    batch_count = batches_count;
    SetTraceLogLevel(LOG_WARNING);
    SetConfigFlags(FLAG_MSAA_4X_HINT);
    InitWindow((int)width, (int)height, "Silk renderer comparison");
    ready = IsWindowReady();
    if (!ready) {
        sl_render_dispose();
        return false;
    }
    SetTargetFPS(0);
    return true;
}
float *sl_render_vertices(void)
{
    return vertices;
}
float *sl_render_poses(void)
{
    return poses;
}
uint32_t *sl_render_batches(void)
{
    return batches;
}

bool sl_render_draw(float scale, float offset_x, float offset_y)
{
    if (!ready || !isfinite(scale) || scale <= 0.0f || !isfinite(offset_x) ||
        !isfinite(offset_y)) {
        return false;
    }
    /* Validate ranges before submitting any drawing. These bounded checks are
     * included in CPU submission, not hidden in a separate timing region. */
    for (uint32_t i = 0u; i < batch_count; ++i) {
        const uint32_t *b = batches + 5u * i;
        if (b[0] > instance_count || b[1] > instance_count - b[0] ||
            b[2] > vertex_count || b[3] > vertex_count - b[2] ||
            b[3] % 3u != 0u || b[4] > 1u) {
            return false;
        }
    }
    BeginDrawing();
    ClearBackground((Color){ 16u, 24u, 32u, 255u });
    rlDisableBackfaceCulling();
    rlSetTexture(rlGetTextureIdDefault());
    rlBegin(RL_TRIANGLES);
    rlTexCoord2f(0.5f, 0.5f);
    rlNormal3f(0.0f, 0.0f, 1.0f);
    for (uint32_t i = 0u; i < batch_count; ++i) {
        const uint32_t *b = batches + 5u * i;
        if (b[4] == 0u) {
            rlColor4ub(113u, 146u, 165u, 255u);
        } else {
            rlColor4ub(235u, 189u, 112u, 255u);
        }
        for (uint32_t j = b[0]; j < b[0] + b[1]; ++j) {
            const float *p = poses + 4u * j;
            if (overlay_colors != NULL) {
                overlay_color(overlay_colors[j]);
            }
            for (uint32_t v = b[2]; v < b[2] + b[3]; ++v) {
                const float x = vertices[2u * v];
                const float y = vertices[2u * v + 1u];
                rlVertex2f(offset_x + scale * (p[0] + p[2] * x - p[3] * y),
                           offset_y - scale * (p[1] + p[3] * x + p[2] * y));
            }
        }
    }
    rlEnd();
    if (overlay_lines != NULL) {
        overlay_draw();
    }
    rlSetTexture(0u);
    EndDrawing();
    return true;
}
void sl_render_dispose(void)
{
    if (ready) {
        CloseWindow();
    }
    free(overlay_lines);
    free(overlay_markers);
    free(overlay_colors);
    overlay_lines = NULL;
    overlay_markers = NULL;
    overlay_colors = NULL;
    line_capacity = 0u;
    marker_capacity = 0u;
    line_count = 0u;
    marker_count = 0u;
    free(vertices);
    free(poses);
    free(batches);
    vertices = NULL;
    poses = NULL;
    batches = NULL;
    instance_count = 0u;
    vertex_count = 0u;
    batch_count = 0u;
    ready = false;
}

bool sl_render_overlay_init(uint32_t lines, uint32_t markers)
{
    /* Derived from the existing maximum B/C/J capacities, never dynamic growth.
     */
    if (!ready || overlay_lines != NULL || lines == 0u || markers == 0u ||
        lines > 4u * 65536u + 2u * 262144u + 65536u + 6u ||
        markers > 2u * 262144u + 2u * 65536u + 2u) {
        return false;
    }
    overlay_lines = calloc((size_t)lines * 5u, sizeof(float));
    overlay_markers = calloc((size_t)markers * 3u, sizeof(float));
    overlay_colors = calloc(instance_count, sizeof(float));
    if (overlay_lines == NULL || overlay_markers == NULL ||
        overlay_colors == NULL) {
        free(overlay_lines);
        free(overlay_markers);
        free(overlay_colors);
        overlay_lines = NULL;
        overlay_markers = NULL;
        overlay_colors = NULL;
        return false;
    }
    line_capacity = lines;
    marker_capacity = markers;
    return true;
}
float *sl_render_overlay_lines(void)
{
    return overlay_lines;
}
float *sl_render_overlay_markers(void)
{
    return overlay_markers;
}
float *sl_render_overlay_colors(void)
{
    return overlay_colors;
}
float *sl_render_overlay_mesh(void)
{
    return overlay_mesh;
}
uint32_t *sl_render_overlay_palette(void)
{
    return overlay_palette;
}
static bool color_valid(float code)
{
    return isfinite(code) && code >= 0.0f && code < 16.0f &&
           floorf(code) == code;
}
bool sl_render_overlay_counts(uint32_t lines, uint32_t markers)
{
    if (overlay_lines == NULL || lines > line_capacity ||
        markers > marker_capacity) {
        return false;
    }
    for (uint32_t i = 0u; i < 48u; ++i) {
        if (overlay_palette[i] > 255u) {
            return false;
        }
    }
    for (uint32_t i = 0u; i < 180u; ++i) {
        if (!isfinite(overlay_mesh[i]) || overlay_mesh[i] < -1.0f ||
            overlay_mesh[i] > 1.0f) {
            return false;
        }
    }
    for (uint32_t i = 0u; i < instance_count; ++i) {
        if (!color_valid(overlay_colors[i])) {
            return false;
        }
    }
    for (uint32_t i = 0u; i < lines; ++i) {
        const float *p = overlay_lines + 5u * i;
        for (uint32_t j = 0u; j < 4u; ++j) {
            if (!isfinite(p[j])) {
                return false;
            }
        }
        const float dx = p[2] - p[0], dy = p[3] - p[1];
        if (!isfinite(dx * dx + dy * dy) || !color_valid(p[4])) {
            return false;
        }
    }
    for (uint32_t i = 0u; i < markers; ++i) {
        const float *p = overlay_markers + 3u * i;
        if (!isfinite(p[0]) || !isfinite(p[1]) || !color_valid(p[2])) {
            return false;
        }
    }
    line_count = lines;
    marker_count = markers;
    return true;
}
